lib/Controller/ConsentAdmin.php

```diff
@@ -0,0 +1,418 @@
+
+            // Run AuthProc filters
+            list($targeted_id, $attribute_hash, $attributes) = $this->driveProcessingChain(
+                $idp_metadata,
+                $source,
+                $sp_metadata,
+                $sp_entityid,
+                $attributes,
+                $userid,
+                $hashAttributes,
+                $excludeAttributes
+            );
+
+            // Add a consent (or update if attributes have changed and old consent for SP and IdP exists)
+            if ($action == 'true') {
+                $isStored = $consent_storage->saveConsent($hashed_user_id, $targeted_id, $attribute_hash);
+            } else {
+                if ($action == 'false') {
+                    // Got consent, so this is a request to remove it
+                    $consent_storage->deleteConsent($hashed_user_id, $targeted_id);
+                    $isStored = false;
+                } else {
+                    Logger::info('consentAdmin: unknown action');
+                    $isStored = null;
+                }
+            }
+            $template->data['isStored'] = $isStored;
+            return $template;
+        }
+
+        // Get all consents for user
+        $user_consent_list = $consent_storage->getConsents($hashed_user_id);
+
+        // Parse list of consents
+        $user_consent = [];
+        foreach ($user_consent_list as $c) {
+            $user_consent[$c[0]] = $c[1];
+        }
+
+        // Init template
+        $template = new Template($this->config, 'consentAdmin:consentadmin.twig', 'consentAdmin:consentadmin');
+        $template->getLocalization()->addAttributeDomains();
+
+        $sp_list = [];
+
+        // Process consents for all SP
+        foreach ($all_sp_metadata as $sp_entityid => $sp_values) {
+            // Get metadata for SP
+            $sp_metadata = $metadata->getMetaData($sp_entityid, 'saml20-sp-remote');
+
+            // Run attribute filters
+            list($targeted_id, $attribute_hash, $attributes) = $this->driveProcessingChain(
+                $idp_metadata,
+                $source,
+                $sp_metadata,
+                $sp_entityid,
+                $attributes,
+                $userid,
+                $hashAttributes,
+                $excludeAttributes
+            );
+
+            // Check if consent exists
+            if (array_key_exists($targeted_id, $user_consent)) {
+                $sp_status = "changed";
+                Logger::info('consentAdmin: changed');
+                // Check if consent is valid. (Possible that attributes has changed)
+                if ($user_consent[$targeted_id] == $attribute_hash) {
+                    Logger::info('consentAdmin: ok');
+                    $sp_status = "ok";
+                }
+                // Consent does not exist
+            } else {
+                Logger::info('consentAdmin: none');
+                $sp_status = "none";
+            }
+
+            // Set name of SP
```

Comment:
Replace this and all the elseifs below with simply entityDisplayName method or twig filter